parsec/core/fs/workspacefs/workspacefile.py

```diff
@@ -0,0 +1,179 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2019 Scille SAS
+
+import os
+import re
+from parsec.core.fs.exceptions import FSUnsupportedOperation, FSOSError
+
+
+class WorkspaceFile:
+    """methods of this object should act like trio async file :
+    "https://github.com/python-trio/trio/blob/master/trio/_file_io.py"
+    using transactions"""
+
+    def __init__(self, fd, transactions, mode: str = "r"):
+        self._fd = fd
+        self._transactions = transactions
+        self._open = True
+        self._offset = 0
+        self._firstWrite = True
+        self._fileLenght = -1
+        self._name = None
+        mode.lower()
+        # preventing to open in write and read in same time or write and append or open with no mode
+        if (("w" in mode or "a" in mode) and "r" in mode) or (
+            ("w" in mode and "a" in mode) or re.search("[arw]", mode) is None
+        ):
+            raise ValueError("must have exactly one of create/read/write/append mode")
+        # preventing to open with non-existant mode
+        elif re.search("[^arwb]", mode) is not None:
+            raise ValueError("invalid mode: ", mode)
```

Comment:
It seems like this logic could be simplified, especially the regex part. Why not something like this:
```python
# Read + append is invalid
if "r" in mode and "a" in mode:
    raise ValueError("...")
# Read + write mode is invalid
if "r" in mode and "w" in mode:
    raise ValueError("...")
# Write + append mode is invalid
if "w" in mode and "a" in mode:
    raise ValueError("...")
# No mode selected
if not any(c in mode for c in "rwa"):
    raise ValueError("...")
```
If you find it too verbose, this should also work:
```python
# Exactly one of `r`, `w`, and `a` should be provided
if sum(c in mode for c in "rwa") == 1:
    raise ValueError(...)
```

Reply:
The regex seems usefull to me because I try to catch all the other characters to raise my error. The sum method for checking that the file have been opened with exactly one correct mode is really usefull though.